Residual Block Error

Started using the [cartographer_turtlebot](https://github.com/googlecartographer/cartographer_turtlebot) package in simulation, and I've been running into this error.  It occurs after a random amount of time ranging from a few seconds after starting the cartographer_node to several minutes- the only changes I've made have been remapping a few topic names (mainly odometry and the laser scanner), removing a line launching turtlebot's minimal bringup and a line launching the urg_node (since I'm using my own version of turtlebot) and disabling the imu (changes have been across [this](https://github.com/googlecartographer/cartographer_turtlebot/blob/master/cartographer_turtlebot/configuration_files/turtlebot_urg_lidar_2d.lua) file and [this](https://github.com/googlecartographer/cartographer_turtlebot/blob/master/cartographer_turtlebot/launch/turtlebot.launch) file).  The file I've been launching from is [here](https://github.com/googlecartographer/cartographer_turtlebot/blob/master/cartographer_turtlebot/launch/turtlebot_urg_lidar_2d.launch).  The error I've been getting is shown below:

```
[ WARN] [1491490574.176173648, 624.161000000]: W0406 10:56:14.000000  8798 residual_block.cc:131] 

Error in evaluating the ResidualBlock.

There are two possible reasons. Either the CostFunction did not evaluate and fill all    
residual and jacobians that were requested or there was a non-finite value (nan/infinite)
generated during the or jacobian computation. 

Residual Block size: 1 parameter blocks x 184 residuals

For each parameter block, the value of the parameters are printed in the first column   
and the value of the jacobian under the corresponding residual. If a ParameterBlock was 
held constant then the corresponding jacobian is printed as 'Not Computed'. If an entry 
of the Jacobian/residual array was requested but was not written to by user code, it is 
indicated by 'Uninitialized'. This is an error. Residuals or Jacobian values evaluating 
to Inf or NaN is also an error.  

Residuals:              nan          nan          nan          nan          nan          nan          nan          nan          nan          nan          nan          nan          nan          nan          nan          nan          nan          nan          nan          nan          nan          nan          nan          nan          nan          nan          nan          nan          nan          nan          nan          nan          nan          nan          nan          nan          nan          nan          nan          nan          nan          nan          nan          nan          nan          nan          nan          nan          nan          nan          nan          nan          nan          nan          nan          nan          nan          nan          nan          nan          nan          nan          nan          nan          nan          nan          nan          nan          nan          nan          nan          nan          nan          nan          nan          nan          nan          nan          nan          nan          nan          nan          nan          nan          nan          nan          nan          nan          nan          nan          nan          nan          nan          nan          nan          nan          nan          nan          nan          nan          nan          nan          nan          nan          nan          nan          nan          nan          nan          nan          nan          nan          nan          nan          nan          nan          nan          nan          nan          nan          nan          nan          nan          nan          nan          nan          nan          nan          nan          nan          nan          nan          nan          nan          nan          nan          nan          nan          nan          nan          nan          nan          nan          nan          nan          nan          nan          nan          nan          nan          nan          nan          nan          nan          nan          nan          nan          nan          nan          nan          nan          nan          nan          nan          nan          nan          nan          nan          nan          nan          nan          nan          nan          nan          nan          nan          nan          nan          nan          nan          nan          nan          nan          nan 

Parameter Block 0, size: 3

         nan |         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan 
         nan |         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan 
  -0.0744971 |         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan         -nan 


CHOLMOD error: invalid xtype
F0406 10:56:14.176298  8798 covariance_impl.cc:652] Check failed: 'permutation' Must be non NULL 
[FATAL] [1491490574.176547018, 624.161000000]: F0406 10:56:14.000000  8798 covariance_impl.cc:652] Check failed: 'permutation' Must be non NULL 
*** Check failure stack trace: ***
    @     0x7ffff7bb35cd  google::LogMessage::Fail()
    @     0x7ffff7bb5433  google::LogMessage::SendToLog()
    @     0x7ffff7bb315b  google::LogMessage::Flush()
    @     0x7ffff7bb5e1e  google::LogMessageFatal::~LogMessageFatal()
    @           0x6546e1  ceres::internal::CovarianceImpl::ComputeCovarianceValuesUsingSuiteSparseQR()
    @           0x65b0a5  ceres::internal::CovarianceImpl::ComputeCovarianceValues()
    @           0x65b22f  ceres::internal::CovarianceImpl::Compute()
    @           0x5a6918  cartographer::mapping_2d::scan_matching::CeresScanMatcher::Match()
    @           0x5ae390  cartographer::mapping_2d::LocalTrajectoryBuilder::ScanMatch()
    @           0x5ae94e  cartographer::mapping_2d::LocalTrajectoryBuilder::AddHorizontalRangeData()
    @           0x5acfbc  cartographer::mapping_2d::GlobalTrajectoryBuilder::AddRangefinderData()
    @           0x5377f9  cartographer::mapping::CollatedTrajectoryBuilder::HandleCollatedSensorData()
    @           0x53822e  _ZNSt17_Function_handlerIFvRKNSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEESt10unique_ptrIN12cartographer6sensor4DataESt14default_deleteISB_EEEZNS9_7mapping25CollatedTrajectoryBuilderC4EPNSA_8CollatorEiRKSt13unordered_setIS5_St4hashIS5_ESt8equal_toIS5_ESaIS5_EES8_INSG_32GlobalTrajectoryBuilderInterfaceESC_IST_EEEUlS7_SE_E_E9_M_invokeERKSt9_Any_dataS7_OSE_
    @           0x61327a  _ZNSt17_Function_handlerIFvSt10unique_ptrIN12cartograp9_Any_dataOS6_
    @           0x6171f9  cartographer::sensor::OrderedMultiQueue::Dispatch()
    @           0x61879f  cartographer::sensor::OrderedMultiQueue::Add()
    @           0x613574  cartographer::sensor::Collator::AddSensorData()
    @           0x536948  cartographer::mapping::CollatedTrajectoryBuilder::AddS
    @           0x509f5b  cartographer_ros::SensorBridge::HandleRangefinder()
    @           0x50ab9a  cartographer_ros::SensorBridge::HandleLaserScanMessage
    @           0x4efb2b  _ZN5boost6detail8function26void_function_obj_invoker1I
    @           0x4f6c4e  boost::detail::function::void_function_obj_invoker1<>:
    @           0x500e39  ros::SubscriptionCallbackHelperT<>::call()
    @     0x7ffff4d1c5cd  ros::SubscriptionQueue::call()
    @     0x7ffff4cc6cf0  ros::CallbackQueue::callOneCB()
    @     0x7ffff4cc80f3  ros::CallbackQueue::callAvailable()
    @     0x7ffff4d20691  ros::SingleThreadedSpinner::spin()
    @     0x7ffff4d0572b  ros::spin()
    @           0x4f0f6e  cartographer_ros::(anonymous namespace)::Run()
    @           0x4ed6a4  main
    @     0x7ffff371f830  __libc_start_main
    @           0x4ef129  _start

Thread 1 "cartographer_no" received signal SIGABRT, Aborted.
0x00007ffff3734428 in __GI_raise (sig=sig@entry=6)
    at ../sysdeps/unix/sysv/linux/raise.c:54
54      ../sysdeps/unix/sysv/linux/raise.c: No such file or directory.
```

It's also worth noting that this error occurs regardless of if the robot stands still or moves- and while moving, does correctly generate a map.  I've tested running the 2D Lidar Demo (listed [here](https://google-cartographer-ros-for-turtlebots.readthedocs.io/en/latest/#running-the-demo)) with imu disabled and no other changes, in order to make sure the disabled imu wasn't causing the issue.  It ran without error under those circumstances.  I'm using Ubuntu 16.04 and ROS Kinetic.

Thanks for the replies- I tried setting the CMakeLists.txt to use EigenSparse, but when I did I got the following error during runtime:

```
E0419 13:57:37.598997  8287 covariance_impl.cc:548] SuiteSparse is required to use the SUITE_SPARSE_QR algorithm.
[ERROR] [1492624657.599212015, 26.529000000]: E0419 13:57:37.000000  8287 covariance_impl.cc:548] SuiteSparse is required to use the SUITE_SPARSE_QR algorithm.
F0419 13:57:37.599295  8287 ceres_scan_matcher.cc:108] Check failed: covariance_computer.Compute(covariance_blocks, &problem) 
[FATAL] [1492624657.599386618, 26.529000000]: F0419 13:57:37.000000  8287 ceres_scan_matcher.cc:108] Check failed: covariance_computer.Compute(covariance_blocks, &problem)
```

The code I added to the CMake file is below- I added it after the options setup.

```
update_cache_variable(SUITESPARSE OFF)
update_cache_variable(EIGENSPARSE ON)
```

I'm guessing I'm setting it up incorrectly

--- a/cartographer/mapping_2d/local_trajectory_builder.cc
+++ b/cartographer/mapping_2d/local_trajectory_builder.cc
@@ -201,7 +201,8 @@ LocalTrajectoryBuilder::AddHorizontalRangeData(
   }
 
   // Improve the velocity estimate.
-  if (last_scan_match_time_ > common::Time::min()) {
+  if (last_scan_match_time_ > common::Time::min() &&
+      time > last_scan_match_time_) {
     const double delta_t = common::ToSeconds(time - last_scan_match_time_);
     velocity_estimate_ += (pose_estimate_.translation().head<2>() -
                            model_prediction.translation().head<2>()) /

